mesonbuild/compilers/compilers.py

```diff
@@ -637,9 +637,13 @@ def __repr__(self):
     def can_compile(self, src):
         if hasattr(src, 'fname'):
             src = src.fname
-        suffix = os.path.splitext(src)[1].lower()
+        # We try first in case sensitive mode detect .C as C++ code instead of C
+        suffix = os.path.splitext(src)[1]
         if suffix and suffix[1:] in self.can_compile_suffixes:
             return True
+        # We keep the old case insensitive behaviour
+        if suffix.lower() and suffix[1:] in self.can_compile_suffixes:
+            return True
```

Comment:
Why is this needed? We have always required the suffix to be lower case, so there's no old case-insensitive behaviour to keep.

Reply:
Proposal: On Linux and BSDs, GCC treats `.C` as a C++ extension, so we should keep the `suffix.lower()` fallback only for other extensions. Looking at the GCC man page, we should also treat `.H` as C++ header for the same reason. The only other extensions that change behaviour in a case-sensitive manner are `.f` and `.F` for fixed-form Fortran and `.s` (asm code) and `.S` (asm code that must be pre-processed).

Reply:
I'm not sure I understand exactly what you want ma to do. 
At least I guess that we agree that we need to do the right language detection from Meson to be able to pass right args?
I know that stupid file systems(case insensitive ones) can make this difficult but it also does for git and many other things.

Reply:
What I mean, is that we should not fallback to detecting `.C` as C sources if a C++ compiler is not found. i.e., we should *not* keep the "old case insensitive behaviour" in that case. We should only maintain case-insensitivity for extensions that *are* case-insensitive.